### PR TITLE
refactor: pull out instanceRecord and issuerStorage into separate files

### PR DESCRIPTION
**Pull request by katelynsills**
Blocked by https://github.com/Agoric/agoric-sdk/pull/2894

This PR gets rid of the `issuerTable` (Note: dapp-svelte-wallet uses it, and rather than allow scope-creep to occur and start refactoring dapp-svelte-wallet too, I moved issuerTable into dapp-svelte wallet). 

This PR also saves on a number of roundtrips to get information about issuers in `contractFacet.js`. Now that `amountMath` is just a library of functions and no longer part of the stored issuer data, we can just take the exported issuer data from the Zoe Service and use that.

Additionally, `E(zoeInstanceAdmin).saveIssuer(issuerP, keyword)` now returns all the issuer data needed to save an issuer to ZCF, so we save a remote call there as well.

### CODE REVIEW

**Comment by erights** [comment]
Any idea why two CI tests are not finishing?

**Comment by katelynsills** [comment]
> Any idea why two CI tests are not finishing?

Hmm, no, that is very strange.

**Review by erights — CHANGES_REQUESTED** [review]
LGTM.

I am frankly unsure what I think of the style of unpacking all the methods from a stateful singleton abstraction and using them as bare functions. It certainly does not violate any rule in my head, and much code (including your's and mine) does this with no problem.

Here, having created these abstractions for managing a current instanceRecord and of an issuerTable, it becomes less clear which stateful singleton we're talking to. Part of the problem is the total number of these methods. The stateful abstractions genuinely are singletons to the code that's calling them. Otherwise the callers would still be using method call syntax to distinguish which instance they're talking to. If they were stateless, the absence of the receiver at the call site wouldn't bother me. But since they're stateful, I'd like to have a better sense at the call site of what state I'm mutating or querying.

I am not confident of this judgement. In fact, it kinda surprises me. If it seems wrong to you, *please* argue with me before merging anything.

**Comment by katelynsills** [comment]
> I am frankly unsure what I think of the style of unpacking all the methods from a stateful singleton abstraction and using them as bare functions. It certainly does not violate any rule in my head, and much code (including your's and mine) does this with no problem.

@erights, if it helps, bare functions aren't the intended end-state of the refactoring. I'm still on the fence about whether this is a good idea or not, but I would like to try to combine the bare functions into objects representing certain authorities. For example, all the authority needed to start a new ZCF vat. I was holding off on finding the right combination until I had done more refactoring of `startInstance`.  But, however it turns out, I would expect the number of bare functions to go down significantly.

File: packages/zoe/src/issuerStorage.js
Comment (katelynsills) [reply packages/zoe/src/issuerStorage.js]: This is not new in this PR, but was in the original issuerTable. I'm just making sure we're consistent. I plan to clean this up in a future PR that renames `mathKind` to `assetKind` and includes it in `displayInfo`

File: packages/zoe/src/issuerStorage.js
Comment (katelynsills) [reply packages/zoe/src/issuerStorage.js]: Whoops, good catch

File: packages/zoe/src/instanceRecordStorage.js
Comment (katelynsills) [reply packages/zoe/src/instanceRecordStorage.js]: Sounds good to me!

File: packages/zoe/src/zoeService/zoe.js
Comment (katelynsills) [reply packages/zoe/src/zoeService/zoe.js]: Yup, I'll hide this behind a function for now, but I expect this to be cleaned up further in a later PR.

File: packages/zoe/src/contractFacet/contractFacet.js
Comment (katelynsills) [reply packages/zoe/src/contractFacet/contractFacet.js:186]: No two functions are always called together, and issuerStorage and instanceRecordStorage at two different levels of abstraction. The issuers are stored at the Zoe level, but the instanceRecord is at the zoeInstanceAdmin level (i.e. different for every instance) so it's not clear that we actually want to combine them. Here's where the different functions are used:

In the ZoeService to store a ZoeMint issuer, `storeIssuerRecord` + `addIssuerToInstanceRecord` + `makeLocalPurse` are used:

```js
        storeIssuerRecord(localIssuerRecord);
        const localPooledPurse = makeLocalPurse(localIssuer, localBrand);
        addIssuerToInstanceRecord(keyword, localIssuerRecord);
```

In the ZoeService when a contract in the ZCF vat saves a new issuer, `storeIssuer` + `addIssuerToInstanceRecord` + `createPurse` are used:

```js
        // checks of keyword done on zcf side
        saveIssuer: async (issuerP, keyword) => {
          const issuerRecord = await storeIssuer(issuerP);
          // AWAIT ///
          createPurse(issuerRecord.issuer, issuerRecord.brand);
          addIssuerToInstanceRecord(keyword, issuerRecord);
          return issuerRecord;
        },
```

In ZCF when a ZCF Mint is created: `storeIssuerRecord` + `addIssuerToInstanceRecord` are used. No purse is created:

```js
      addIssuerToInstanceRecord(keyword, mintyIssuerRecord);
      storeIssuerRecord(mintyIssuerRecord);
```

In ZCF when a contract saves a new issuer: `storeIssuerRecord` + `addIssuerToInstanceRecord` are used. No purse is created.

```js
      saveIssuer: async (issuerP, keyword) => {
        // TODO: The checks of the keyword for uniqueness are
        // duplicated. Assess how waiting on promises to resolve might
        // affect those checks and see if one can be removed.
        assertUniqueKeyword(keyword);
        const record = await E(zoeInstanceAdmin).saveIssuer(issuerP, keyword);
        // AWAIT ///
        storeIssuerRecord(record);
        addIssuerToInstanceRecord(keyword, record);
        return record;
      },
 ```
 
 So it looks like we could potentially combine `storeIssuerRecord` and  `addIssuerToInstanceRecord` but probably only should do it in ZCF. I'd like to push this change to a later PR when the authority usage overall is clearer.

File: packages/zoe/src/instanceRecordStorage.js
Comment (katelynsills) [reply packages/zoe/src/instanceRecordStorage.js]: Fixed

File: packages/zoe/src/instanceRecordStorage.js
Comment (katelynsills) [reply packages/zoe/src/instanceRecordStorage.js:68]: Added comments.

File: packages/zoe/src/issuerStorage.js
Comment (katelynsills) [reply packages/zoe/src/issuerStorage.js:58]: >  If, for example, the new record has a different brand, you'll return the record according to the issuer rather than the brand.

The new record can't have a different brand than the brand that matches the issuer according to mutual agreement. We use `storeIssuerRecord` in 4 cases: 1) ZoeMint records, 2) ZCFMint records, 3) within `storeIssuer`, and 4) storing an issuer record in ZCF that we obtained from Zoe that originally went through `storeIssuer`. In both ZoeMints and ZCFMints we create the issuers and brands ourselves using ERTP directly, so we know they are not malicious and the issuers and brands match. In the last two cases, we go through the brand-issuer-match check in `storeIssuer`. 

The reason why we need the `has` check in lines 32-34 is for the 4th case above, in which we store an issuer record in ZCF that we obtained from Zoe. `init` errors if the key is already present, and because an issuer can be used more than once in the same contract, we need to make sure we aren't trying to `init` twice. If it's already present, we do not need to add it again in ZCF.

I just tried commenting out lines 32-34 and I get 16 tests failed. Could it have been possible that you didn't `build`?

I'll admit that this could be more obvious rather than requiring an analysis, but I think it's sound as is. I can copy this write-up into a comment if that'd be helpful.

Comment (katelynsills) [reply packages/zoe/src/issuerStorage.js:58]: Added a comment that is a cleaned-up version of ^

Comment (erights) [reply packages/zoe/src/issuerStorage.js:58]: > I just tried commenting out lines 32-34 and I get 16 tests failed. Could it have been possible that you didn't build?

Ah. yes.

File: packages/zoe/src/contractFacet/contractFacet.js
Comment (erights) [reply packages/zoe/src/contractFacet/contractFacet.js:186]: >  I'd like to push this change to a later PR when the authority usage overall is clearer.

yes

**Comment by erights** [comment]
> @erights, if it helps, bare functions aren't the intended end-state of the refactoring. I'm still on the fence about whether this is a good idea or not, but I would like to try to combine the bare functions into objects representing certain authorities. For example, all the authority needed to start a new ZCF vat. I was holding off on finding the right combination until I had done more refactoring of startInstance. But, however it turns out, I would expect the number of bare functions to go down significantly.

This sounds great! Looking forward to seeing that experiment. Knowing that this PR is an intermediate state towards that experiment makes me much happier. But still, having the intermediate state still keep some grouping of these functions back into methods shouldn't hurt that eventual refactoring, and will make the result of this PR much more understandable.

File: packages/zoe/src/issuerStorage.js
Comment (erights) [reply packages/zoe/src/issuerStorage.js:58]: I like the new comment. Resolving.

File: packages/zoe/src/issuerStorage.js
Comment (erights) [reply packages/zoe/src/issuerStorage.js]: I like the new function that localizes this concern. Resolving.

**Review by erights — APPROVED** [review]
LGTM!

Given the eventual refactoring, it's up to you whether to do something now to group the bare function, or just leave that to the later PR. But I would prefer that they stay grouped, even if according to the wrong rationale.

**Comment by katelynsills** [comment]
> LGTM!
> 
> Given the eventual refactoring, it's up to you whether to do something now to group the bare function, or just leave that to the later PR. But I would prefer that they stay grouped, even if according to the wrong rationale.

Ok, thanks! I'd really like to leave it for the later PR, as I have a few PRs/pre-PR-branches stacked on this one, and I think things will change considerably.

**Pull request merged**
